File: benchmarks/trie.c

#include <stdbool.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct TrieNode {
  struct TrieNode *children[26];
  bool is_word;
} TrieNode;

TrieNode *create_node() {
  TrieNode *node = (TrieNode *)malloc(sizeof(TrieNode));
  for (int i = 0; i < 26; i++) {
    node->children[i] = NULL;
  }
  node->is_word = false;
  return node;
}

void insert(TrieNode *root, const char *word) {
  TrieNode *curr = root;
  for (int i = 0; word[i] != '\0'; i++) {
    int idx = word[i] - 'a';
    if (curr->children[idx] == NULL) {
      curr->children[idx] = create_node();
    }
    curr = curr->children[idx];
  }
  curr->is_word = true;
}

bool search(TrieNode *root, const char *word) {
  TrieNode *curr = root;
  for (int i = 0; word[i] != '\0'; i++) {
    int idx = word[i] - 'a';
    if (curr->children[idx] == NULL) {
      return false;
    }
    curr = curr->children[idx];
  }
  return curr->is_word;
}

int main() {
  TrieNode *root = create_node();

  char word[6];
  word[5] = '\0';

  printf("Inserting 700k words...\n");
  for (int i = 0; i < 700000; i++) {
    word[0] = (i % 26) + 97;
    word[1] = ((i / 26) % 26) + 97;
    word[2] = ((i / 676) % 26) + 97;
    word[3] = ((i / 17576) % 26) + 97;
    word[4] = (i % 7) + 97;
    insert(root, word);
  }

  printf("Searching 700k words...\n");
  int found = 0;
  for (int i = 0; i < 700000; i++) {
    word[0] = (i % 26) + 97;
    word[1] = ((i / 26) % 26) + 97;
    word[2] = ((i / 676) % 26) + 97;
    word[3] = ((i / 17576) % 26) + 97;
    word[4] = (i % 7) + 97;
    if (search(root, word)) {
      found++;
    }
  }
  printf("Found: %d\n", found);

  return 0;
}

File: benchmarks/trie.rs

#[derive(Default)]
struct TrieNode {
    children: [Option<Box<TrieNode>>; 26],
    is_word: bool,
}

impl TrieNode {
    fn new() -> Self {
        TrieNode {
            children: Default::default(),
            is_word: false,
        }
    }
}

fn insert(root: &mut TrieNode, word: &[u8]) {
    let mut curr = root;
    for &c in word {
        let idx = (c - b'a') as usize;
        if curr.children[idx].is_none() {
            curr.children[idx] = Some(Box::new(TrieNode::new()));
        }
        curr = curr.children[idx].as_mut().unwrap();
    }
    curr.is_word = true;
}

fn search(root: &TrieNode, word: &[u8]) -> bool {
    let mut curr = root;
    for &c in word {
        let idx = (c - b'a') as usize;
        if let Some(ref node) = curr.children[idx] {
            curr = node;
        } else {
            return false;
        }
    }
    curr.is_word
}

fn main() {
    let mut root = TrieNode::new();
    let mut word = [0u8; 5];

    println!("Inserting 700k words...");
    for i in 0..700000u32 {
        word[0] = ((i % 26) + 97) as u8;
        word[1] = (((i / 26) % 26) + 97) as u8;
        word[2] = (((i / 676) % 26) + 97) as u8;
        word[3] = (((i / 17576) % 26) + 97) as u8;
        word[4] = ((i % 7) + 97) as u8;
        insert(&mut root, &word);
    }

    println!("Searching 700k words...");
    let mut found = 0i64;
    for i in 0..700000u32 {
        word[0] = ((i % 26) + 97) as u8;
        word[1] = (((i / 26) % 26) + 97) as u8;
        word[2] = (((i / 676) % 26) + 97) as u8;
        word[3] = (((i / 17576) % 26) + 97) as u8;
        word[4] = ((i % 7) + 97) as u8;
        if search(&root, &word) {
            found += 1;
        }
    }
    println!("Found: {}", found);
}
